frame/fee-market/src/s2s/payment.rs

```diff
@@ -139,47 +139,71 @@ where
+			match order.lowest_and_highest_fee() {
+				(Some(lowest_fee), Some(message_fee)) => {
+					let message_reward;
+					let confirm_reward;
 
-				// 60% * lowest fee => assigned_relayers_rewards
-				let assigned_relayers_reward = T::AssignedRelayersRewardRatio::get() * lowest_fee;
-				assigned_relayers_rewards
-					.entry(who)
-					.and_modify(|r| *r = r.saturating_add(assigned_relayers_reward))
-					.or_insert(assigned_relayers_reward);
+					if let Some(who) = order.required_delivery_relayer_for_time(order_confirm_time)
+					{
+						// message fee - lowest fee => treasury
+						let treasury_reward = message_fee.saturating_sub(lowest_fee);
+						treasury_total_rewards =
+							treasury_total_rewards.saturating_add(treasury_reward);
 
-				let bridger_relayers_reward = lowest_fee.saturating_sub(assigned_relayers_reward);
-				// 80% * (1 - 60%) * lowest_fee => message relayer
-				message_reward = T::MessageRelayersRewardRatio::get() * bridger_relayers_reward;
-				// 20% * (1 - 60%) * lowest_fee => confirm relayer
-				confirm_reward = T::ConfirmRelayersRewardRatio::get() * bridger_relayers_reward;
-			} else {
-				// The message is delivered by common relayer instead of order assigned relayers, all assigned relayers of this order should be punished.
-				let slashed_reward = slash_assigned_relayers::<T>(order, relayer_fund_account);
+						// 60% * lowest fee => assigned_relayers_rewards
+						let assigned_relayers_reward =
+							T::AssignedRelayersRewardRatio::get() * lowest_fee;
+						assigned_relayers_rewards
+							.entry(who)
+							.and_modify(|r| *r = r.saturating_add(assigned_relayers_reward))
+							.or_insert(assigned_relayers_reward);
 
-				// 80% total slash => confirm relayer
-				message_reward = T::MessageRelayersRewardRatio::get() * slashed_reward;
-				// 20% total slash => confirm relayer
-				confirm_reward = T::ConfirmRelayersRewardRatio::get() * slashed_reward;
-			}
+						let bridger_relayers_reward =
+							lowest_fee.saturating_sub(assigned_relayers_reward);
+						// 80% * (1 - 60%) * lowest_fee => message relayer
+						message_reward =
+							T::MessageRelayersRewardRatio::get() * bridger_relayers_reward;
+						// 20% * (1 - 60%) * lowest_fee => confirm relayer
+						confirm_reward =
+							T::ConfirmRelayersRewardRatio::get() * bridger_relayers_reward;
+					} else {
+						// The message is delivered by common relayer instead of order assigned relayers, all assigned relayers of this order should be slash.
+						// Total slash = message fee + assigned_relayers slash
+						// For each assigned relayer, slash CollateralPerOrder per order
+						let mut total_slash = message_fee;
+
+						let mut assigned_relayers_slash = RingBalance::<T>::zero();
+						for assigned_relayer in order.relayers_slice() {
+							let slashed = do_slash::<T>(
+								&assigned_relayer.id,
+								relayer_fund_account,
+								order.locked_collateral,
```

Comment:
Oh, we are going to slash all the collateral in the order at once, not auction model?

I thought it is linear increasing till max value which equals order.locked_collateral.